src/main/java/org/elasticsearch/common/util/concurrent/ReleasableLock.java

```diff
@@ -30,17 +30,48 @@
 public class ReleasableLock implements Releasable {
     private final Lock lock;
 
+    /* a per thread boolean indicating the lock is held by it. only works when assertions are enabled */
+    private final ThreadLocal<Boolean> holdingThreads;
+
     public ReleasableLock(Lock lock) {
         this.lock = lock;
+        boolean useHoldingThreads = false;
+        assert (useHoldingThreads = true);
+        if (useHoldingThreads) {
+            holdingThreads = new ThreadLocal<>();
+        } else {
+            holdingThreads = null;
+        }
     }
 
     @Override
     public void close() {
         lock.unlock();
+        assert removeCurrentThread();
     }
 
+
     public ReleasableLock acquire() throws EngineException {
         lock.lock();
+        assert addCurrentThread();
         return this;
     }
+
+    private boolean addCurrentThread() {
+        holdingThreads.set(true);
+        return true;
+    }
+
+    private boolean removeCurrentThread() {
+        holdingThreads.remove();
+        return true;
+    }
+
+    public Boolean isHeldByCurrentThread() {
```

Comment:
the whole point here is to be able to ask a lock if it's held by the current thread so it has to be public. I made it a Boolean because it may not know (if assertions are not enabled).  can change to `assertIsHeldByCurrentThread` in the hope it makes it clearer, but then what do we return when the info is not available , true?